src/Npgsql/NpgsqlCommand.cs

```diff
@@ -1147,9 +1147,37 @@ async ValueTask<NpgsqlDataReader> ExecuteReaderAsync(CommandBehavior behavior, b
                 if (conn.TryGetBoundConnector(out var connector))
                 {
                     connector.StartUserAction(this);
+
+                    CancellationTokenRegistration? registration = null;
+
                     try
                     {
-                        using var _ = cancellationToken.Register(cmd => ((NpgsqlCommand)cmd!).Cancel(), this);
+                        var finalCt = CancellationToken.None;
+
+                        if (cancellationToken.CanBeCanceled)
+                        {
+                            if (connector.CommandCts.IsCancellationRequested)
+                            {
+                                connector.CommandCts.Dispose();
+                                connector.CommandCts = new CancellationTokenSource();
+                            }
+
+                            registration = cancellationToken.Register(o =>
+                            {
+                                (var cmd, var cn) = (Tuple<NpgsqlCommand, NpgsqlConnector>)o!;
+                                try
+                                {
+                                    cmd.Cancel(true);
+                                    if (cn.Settings.CancellationTimeout > 0)
+                                        cn.CommandCts.CancelAfter(cn.Settings.CancellationTimeout * 1000);
+                                }
+                                catch
+                                {
+                                    cn.CommandCts.Cancel();
+                                }
+                            }, Tuple.Create(this, connector));
```

Comment:
Tuple.Create allocates - we should get the connector again through the command instance instead.

Note that we need to be resilient to various race conditions here; when the user triggers the cancellation, the connection may be being broken etc., so we should guard against that.